handler/src/main/java/io/netty/handler/ssl/ConscryptAlpnSslEngine.java

```diff
@@ -0,0 +1,239 @@
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLHandshakeException;
+import org.conscrypt.OpenSSLEngineImpl;
+
+/**
+ * A {@link JdkSslEngine} that uses the Conscrypt provider or SSL with ALPN.
+ */
+abstract class ConscryptAlpnSslEngine extends JdkSslEngine {
+
+    private static Class engineClass;
+
+    private boolean handshakeComplete;
+
+    /**
+     * Indicates whether or not conscrypt is available on the current system.
+     */
+    static boolean isAvailable() {
+        if (engineClass == null) {
+            try {
+                // Always use bootstrap class loader.
+                engineClass = Class.forName("org.conscrypt.OpenSSLEngineImpl", true,
+                        ConscryptAlpnSslEngine.class.getClassLoader());
+            } catch (Throwable ignore) {
+                // Conscrypt was not loaded.
+            }
+        }
+        return engineClass != null;
+    }
+
+    static boolean isEngineSupported(SSLEngine engine) {
+        return isAvailable() && engineClass.equals(engine.getClass());
+    }
+
+    static ConscryptAlpnSslEngine newClientEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ClientEngine((OpenSSLEngineImpl) engine, applicationNegotiator);
+    }
+
+    static ConscryptAlpnSslEngine newServerEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ServerEngine((OpenSSLEngineImpl) engine, applicationNegotiator);
+    }
+
+    @Override
+    public final void beginHandshake() throws SSLException {
+        super.beginHandshake();
+        updateHandshakeStatus();
+    }
+
+    @Override
+    public final SSLEngineResult wrap(ByteBuffer byteBuffer, ByteBuffer byteBuffer2) throws SSLException {
+        return updateHandshakeStatus(super.wrap(byteBuffer, byteBuffer2));
+    }
+
+    @Override
+    public final SSLEngineResult wrap(ByteBuffer[] byteBuffers, ByteBuffer byteBuffer) throws SSLException {
+        return updateHandshakeStatus(super.wrap(byteBuffers, byteBuffer));
+    }
+
+    @Override
+    public final SSLEngineResult wrap(ByteBuffer[] byteBuffers, int i, int i2, ByteBuffer byteBuffer)
+            throws SSLException {
+        return updateHandshakeStatus(super.wrap(byteBuffers, i, i2, byteBuffer));
+    }
+
+    @Override
+    public final SSLEngineResult unwrap(ByteBuffer byteBuffer, ByteBuffer byteBuffer2) throws SSLException {
+        return updateHandshakeStatus(super.unwrap(byteBuffer, byteBuffer2));
+    }
+
+    @Override
+    public final SSLEngineResult unwrap(ByteBuffer byteBuffer, ByteBuffer[] byteBuffers) throws SSLException {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)